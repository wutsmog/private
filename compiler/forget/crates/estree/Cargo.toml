[package]
name = "estree"
version = "0.1.0"
edition = "2021"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
bumpalo = { version = "3.13.0", features = ["boxed", "collections"] }
elsa = "1.8.1"
insta = { version = "1.30.0", features = ["glob"] }
serde = { version = "1.0.164", features = ["derive"] }
serde_json = "1.0.99"
static_assertions = "1.1.0"

[build-dependencies]
estree-codegen = { path = "../estree-codegen" }