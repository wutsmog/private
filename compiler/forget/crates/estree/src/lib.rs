mod binding;
mod generated;
mod generated_extensions;
mod js_value;
mod range;

pub use binding::{Binding, BindingId};
pub use generated::*;
pub use js_value::JsValue;
pub use range::SourceRange;

#[cfg(test)]
mod tests {
    use super::*;
    use insta::{assert_snapshot, glob};
    use serde_json;

    #[test]
    fn fixtures() {
        glob!("fixtures/**.json", |path| {
            println!("{:?}", path);
            let input = std::fs::read_to_string(path).unwrap();
            let ast: Program = serde_json::from_str(&input).unwrap();
            let serialized = serde_json::to_string_pretty(&ast).unwrap();
            assert_snapshot!(format!("Input:\n{input}\n\nOutput:\n{serialized}"));
        });
    }
}
