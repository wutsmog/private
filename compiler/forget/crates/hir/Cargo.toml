[package]
name = "hir"
version = "0.1.0"
edition = "2021"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
bumpalo = { version = "3.13.0", features = ["boxed", "collections"] }
estree = { path = "../estree" }
indexmap = "2.0.0"
serde = "1.0.164"
utils = { path = "../utils" }
