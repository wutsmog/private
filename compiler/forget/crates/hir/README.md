# HIR

This crate defines the core data structures that Forget uses to represent and compile input programs. 
