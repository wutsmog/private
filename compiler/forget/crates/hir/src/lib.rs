mod basic_block;
mod environment;
mod features;
mod function;
mod id_types;
mod instruction;
mod registry;
mod terminal;
mod types;

pub use basic_block::*;
pub use environment::*;
pub use features::*;
pub use function::*;
pub use id_types::*;
pub use instruction::*;
pub use registry::Registry;
pub use terminal::*;
pub use types::*;
