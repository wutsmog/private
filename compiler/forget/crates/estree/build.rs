use estree_codegen::estree;

// Example custom build script.
fn main() {
    // Re-run if the codegen files change
    println!("cargo:rerun-if-changed=../estree-codegen/src/codegen.rs");
    println!("cargo:rerun-if-changed=../estree-codegen/src/lib.rs");
    println!("cargo:rerun-if-changed=../estree-codegen/src/ecmascript.json");
    println!("cargo:rerun-if-changed=../estree-codegen");

    let src = estree();
    std::fs::write("src/generated.rs", src).unwrap();
}
