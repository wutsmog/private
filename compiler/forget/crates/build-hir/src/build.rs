use bumpalo::collections::{String, Vec};
use estree::{
    AssignmentTarget, BinaryExpression, BlockStatement, Expression, ForInit, ForStatement,
    FunctionDeclaration, IfStatement, JsValue, Literal, Pattern, Statement,
    VariableDeclarationKind,
};
use hir::{
    ArrayElement, BlockKind, BranchTerminal, Environment, ForTerminal, Function, GotoKind,
    IdentifierOperand, InstrIx, InstructionKind, InstructionValue, LValue, LoadGlobal, LoadLocal,
    Operand, PrimitiveValue, TerminalValue,
};

use crate::{
    builder::{Binding, Builder, LoopScope},
    error::DiagnosticError,
    BuildDiagnostic, ErrorSeverity,
};

/// Converts a React function in ESTree format into HIR. Returns the HIR
/// if it was constructed sucessfully, otherwise a list of diagnostics
/// if the input could be not be converted to HIR.
///
/// Failures generally include nonsensical input (`delete 1`) or syntax
/// that is not yet supported.
pub fn build<'a>(
    env: &'a Environment<'a>,
    fun: FunctionDeclaration,
) -> Result<&'a mut Function<'a>, BuildDiagnostic> {
    let mut builder = Builder::new(env);

    match fun.function.body {
        Some(estree::FunctionBody::BlockStatement(body)) => {
            lower_block_statement(env, &mut builder, *body)?
        }
        Some(estree::FunctionBody::Expression(body)) => {
            lower_expression(env, &mut builder, body)?;
        }
        None => {
            return Err(BuildDiagnostic::new(
                DiagnosticError::EmptyFunction,
                ErrorSeverity::InvalidSyntax,
                fun.range,
            ));
        }
    }

    let mut params = Vec::with_capacity_in(fun.function.params.len(), &env.allocator);
    for param in fun.function.params {
        match param {
            Pattern::Identifier(param) => {
                let identifier = lower_identifier_for_assignment(
                    env,
                    &mut builder,
                    InstructionKind::Let,
                    *param,
                )?;
                params.push(identifier);
            }
        }
    }

    // In case the function did not explicitly return, terminate the final
    // block with an explicit `return undefined`. If the function *did* return,
    // this will be unreachable and get pruned later.
    let implicit_return_value = builder.push(InstructionValue::Primitive(hir::Primitive {
        value: PrimitiveValue::Undefined,
    }));
    builder.terminate(
        TerminalValue::Return(hir::ReturnTerminal {
            value: Operand {
                ix: implicit_return_value,
                effect: None,
            },
        }),
        hir::BlockKind::Block,
    );

    let body = builder.build()?;
    Ok(env.alloc(Function {
        id: fun
            .function
            .id
            .map(|id| String::from_str_in(&id.name, &env.allocator)),
        body,
        params,
        is_async: fun.function.is_async,
        is_generator: fun.function.is_generator,
    }))
}

fn lower_block_statement<'a>(
    env: &'a Environment<'a>,
    builder: &mut Builder<'a>,
    stmt: BlockStatement,
) -> Result<(), BuildDiagnostic> {
    for stmt in stmt.body {
        lower_statement(env, builder, stmt, None)?;
    }
    Ok(())
}

/// Convert a statement to HIR. This will often result in multiple instructions and blocks
/// being created as statements often describe control flow.
fn lower_statement<'a>(
    env: &'a Environment<'a>,
    builder: &mut Builder<'a>,
    stmt: Statement,
    label: Option<String<'a>>,
) -> Result<(), BuildDiagnostic> {
    match stmt {
        Statement::BlockStatement(stmt) => {
            lower_block_statement(env, builder, *stmt)?;
        }
        Statement::BreakStatement(stmt) => {
            let block = builder.resolve_break(stmt.label.as_ref())?;
            builder.terminate(
                TerminalValue::Goto(hir::GotoTerminal {
                    block,
                    kind: GotoKind::Break,
                }),
                BlockKind::Block,
            );
        }
        Statement::ContinueStatement(stmt) => {
            let block = builder.resolve_continue(stmt.label.as_ref())?;
            builder.terminate(
                TerminalValue::Goto(hir::GotoTerminal {
                    block,
                    kind: GotoKind::Continue,
                }),
                BlockKind::Block,
            );
        }
        Statement::ReturnStatement(stmt) => {
            let ix = match stmt.argument {
                Some(argument) => lower_expression(env, builder, argument)?,
                None => builder.push(InstructionValue::Primitive(hir::Primitive {
                    value: PrimitiveValue::Undefined,
                })),
            };
            builder.terminate(
                TerminalValue::Return(hir::ReturnTerminal {
                    value: Operand { ix, effect: None },
                }),
                BlockKind::Block,
            );
        }
        Statement::ExpressionStatement(stmt) => {
            lower_expression(env, builder, stmt.expression)?;
        }
        Statement::EmptyStatement(_) => {
            // no-op
        }
        Statement::VariableDeclaration(stmt) => {
            let kind = match stmt.kind {
                VariableDeclarationKind::Const => InstructionKind::Const,
                VariableDeclarationKind::Let => InstructionKind::Let,
                VariableDeclarationKind::Var => {
                    return Err(BuildDiagnostic::new(
                        DiagnosticError::VariableDeclarationKindIsVar,
                        ErrorSeverity::Unsupported,
                        stmt.range,
                    ));
                }
            };
            for declaration in stmt.declarations {
                if let Some(init) = declaration.init {
                    let value = lower_expression(env, builder, init)?;
                    lower_assignment(
                        env,
                        builder,
                        kind,
                        AssignmentTarget::Pattern(declaration.id.into()),
                        value,
                    )?;
                } else {
                    if let Pattern::Identifier(id) = declaration.id {
                        // TODO: handle unbound variables
                        let binding = builder.resolve_binding(&id)?;
                        let identifier = match binding {
                            Binding::Local(identifier) => identifier,
                            _ => {
                                return Err(BuildDiagnostic::new(
                                    DiagnosticError::VariableDeclarationBindingIsNonLocal,
                                    ErrorSeverity::Invariant,
                                    id.range,
                                ));
                            }
                        };
                        builder.push(InstructionValue::DeclareLocal(hir::DeclareLocal {
                            lvalue: LValue {
                                identifier: IdentifierOperand {
                                    identifier,
                                    effect: None,
                                },
                                kind,
                            },
                        }));
                    }
                }
            }
        }
        Statement::IfStatement(stmt) => {
            // block for what follows the if statement, though this may
            // not be reachable
            let fallthrough_block = builder.reserve(BlockKind::Block);

            let IfStatement {
                test,
                consequent,
                alternate,
                ..
            } = *stmt;

            let consequent_block = builder.enter(BlockKind::Block, |builder| {
                lower_statement(env, builder, consequent, None)?;
                Ok(TerminalValue::Goto(hir::GotoTerminal {
                    block: fallthrough_block.id,
                    kind: GotoKind::Break,
                }))
            })?;

            let alternate_block = builder.enter(BlockKind::Block, |builder| {
                if let Some(alternate) = alternate {
                    lower_statement(env, builder, alternate, None)?;
                }
                Ok(TerminalValue::Goto(hir::GotoTerminal {
                    block: fallthrough_block.id,
                    kind: GotoKind::Break,
                }))
            })?;

            let test = lower_expression(env, builder, test)?;
            let terminal = TerminalValue::If(hir::IfTerminal {
                test: Operand {
                    ix: test,
                    effect: None,
                },
                consequent: consequent_block,
                alternate: alternate_block,
                fallthrough: Some(fallthrough_block.id),
            });
            builder.terminate_with_fallthrough(terminal, fallthrough_block);
        }
        Statement::ForStatement(stmt) => {
            let ForStatement {
                init,
                test,
                update,
                body,
                ..
            } = *stmt;

            // Block for the loop's test condition
            let test_block = builder.reserve(BlockKind::Loop);

            // Block for code following the loop
            let fallthrough_block = builder.reserve(BlockKind::Block);

            let init_block = builder.enter(BlockKind::Loop, |builder| {
                if let Some(ForInit::VariableDeclaration(decl)) = init {
                    lower_statement(env, builder, Statement::VariableDeclaration(decl), None)?;
                    Ok(TerminalValue::Goto(hir::GotoTerminal {
                        block: test_block.id,
                        kind: GotoKind::Break,
                    }))
                } else {
                    Err(BuildDiagnostic::new(
                        DiagnosticError::ForStatementIsMissingInitializer,
                        ErrorSeverity::Todo,
                        None,
                    ))
                }
            })?;

            let update_block = update
                .map(|update| {
                    builder.enter(BlockKind::Loop, |builder| {
                        lower_expression(env, builder, update)?;
                        Ok(TerminalValue::Goto(hir::GotoTerminal {
                            block: test_block.id,
                            kind: GotoKind::Break,
                        }))
                    })
                })
                .transpose()?;

            let body_block = builder.enter(BlockKind::Block, |builder| {
                let loop_ = LoopScope {
                    label,
                    continue_block: update_block.unwrap_or(test_block.id),
                    break_block: fallthrough_block.id,
                };
                builder.enter_loop(loop_, |builder| {
                    lower_statement(env, builder, body, None)?;
                    Ok(TerminalValue::Goto(hir::GotoTerminal {
                        block: update_block.unwrap_or(test_block.id),
                        kind: GotoKind::Continue,
                    }))
                })
            })?;

            let terminal = TerminalValue::For(ForTerminal {
                body: body_block,
                init: init_block,
                test: test_block.id,
                fallthrough: fallthrough_block.id,
                update: update_block,
            });
            builder.terminate_with_fallthrough(terminal, test_block);

            if let Some(test) = test {
                let test_value = lower_expression(env, builder, test)?;
                let terminal = TerminalValue::Branch(BranchTerminal {
                    test: Operand {
                        ix: test_value,
                        effect: None,
                    },
                    consequent: body_block,
                    alternate: fallthrough_block.id,
                });
                builder.terminate_with_fallthrough(terminal, fallthrough_block);
            } else {
                return Err(BuildDiagnostic::new(
                    DiagnosticError::ForStatementIsMissingTest,
                    ErrorSeverity::Todo,
                    stmt.range,
                ));
            }
        }
        _ => todo!("Lower {stmt:#?}"),
    }
    Ok(())
}

/// Converts an ESTree Expression into an HIR InstructionValue. Note that while only a single
/// InstructionValue is returned, this function is recursive and may cause multiple instructions
/// to be emitted, possibly across multiple basic blocks (in the case of expressions with control
/// flow semenatics such as logical, conditional, and optional expressions).
fn lower_expression<'a>(
    env: &'a Environment<'a>,
    builder: &mut Builder<'a>,
    expr: Expression,
) -> Result<InstrIx, BuildDiagnostic> {
    let value = match expr {
        Expression::Identifier(expr) => {
            // TODO: handle unbound variables
            let binding = builder.resolve_binding(&expr)?;
            match binding {
                Binding::Local(identifier) => {
                    let place = IdentifierOperand {
                        effect: None,
                        identifier,
                    };
                    InstructionValue::LoadLocal(LoadLocal { place })
                }
                Binding::Module(..) | Binding::Global => InstructionValue::LoadGlobal(LoadGlobal {
                    name: String::from_str_in(&expr.name, &env.allocator),
                }),
            }
        }
        Expression::Literal(expr) => InstructionValue::Primitive(hir::Primitive {
            value: lower_primitive(env, builder, *expr),
        }),
        Expression::ArrayExpression(expr) => {
            let mut elements = Vec::with_capacity_in(expr.elements.len(), &env.allocator);
            for expr in expr.elements {
                let element = match expr {
                    Some(estree::ExpressionOrSpread::SpreadElement(expr)) => {
                        Some(ArrayElement::Spread(Operand {
                            ix: lower_expression(env, builder, expr.argument)?,
                            effect: None,
                        }))
                    }
                    Some(estree::ExpressionOrSpread::Expression(expr)) => {
                        Some(ArrayElement::Place(Operand {
                            ix: lower_expression(env, builder, expr)?,
                            effect: None,
                        }))
                    }
                    None => None,
                };
                elements.push(element);
            }
            InstructionValue::Array(hir::Array { elements })
        }

        Expression::AssignmentExpression(expr) => match expr.operator {
            estree::AssignmentOperator::Equals => {
                let right = lower_expression(env, builder, expr.right)?;
                return Ok(lower_assignment(
                    env,
                    builder,
                    InstructionKind::Reassign,
                    expr.left,
                    right,
                )?);
            }
            _ => todo!("lower assignment expr {:#?}", expr),
        },

        Expression::BinaryExpression(expr) => {
            let BinaryExpression {
                left,
                operator,
                right,
                ..
            } = *expr;
            let left = lower_expression(env, builder, left)?;
            let right = lower_expression(env, builder, right)?;
            InstructionValue::Binary(hir::Binary {
                left: Operand {
                    ix: left,
                    effect: None,
                },
                operator,
                right: Operand {
                    ix: right,
                    effect: None,
                },
            })
        }

        _ => todo!("Lower expr {expr:#?}"),
    };
    Ok(builder.push(value))
}

fn lower_assignment<'a>(
    env: &'a Environment<'a>,
    builder: &mut Builder<'a>,
    kind: InstructionKind,
    lvalue: AssignmentTarget,
    value: InstrIx,
) -> Result<InstrIx, BuildDiagnostic> {
    Ok(match lvalue {
        AssignmentTarget::Pattern(lvalue) => match lvalue {
            Pattern::Identifier(lvalue) => {
                let identifier = lower_identifier_for_assignment(env, builder, kind, *lvalue)?;
                builder.push(InstructionValue::StoreLocal(hir::StoreLocal {
                    lvalue: LValue { identifier, kind },
                    value: Operand {
                        ix: value,
                        effect: None,
                    },
                }))
            }
            _ => todo!("lower assignment pattern for {:#?}", lvalue),
        },
        _ => todo!("lower assignment for {:#?}", lvalue),
    })
}

fn lower_identifier_for_assignment<'a>(
    _env: &'a Environment<'a>,
    builder: &mut Builder<'a>,
    _kind: InstructionKind,
    identifier: estree::Identifier,
) -> Result<IdentifierOperand<'a>, BuildDiagnostic> {
    let binding = builder.resolve_binding(&identifier)?;
    match binding {
        Binding::Module(..) | Binding::Global => Err(BuildDiagnostic::new(
            DiagnosticError::ReassignedGlobal,
            ErrorSeverity::InvalidReact,
            identifier.range,
        )),
        Binding::Local(id) => Ok(IdentifierOperand {
            identifier: id,
            effect: None,
        }),
    }
}

/// Converts an ESTree literal into a HIR primitive
fn lower_primitive<'a>(
    env: &'a Environment<'a>,
    _builder: &mut Builder<'a>,
    literal: Literal,
) -> PrimitiveValue<'a> {
    match literal.value {
        JsValue::Bool(bool) => PrimitiveValue::Boolean(bool),
        JsValue::Null => PrimitiveValue::Null,
        JsValue::Number(value) => PrimitiveValue::Number(f64::from(value).into()),
        JsValue::String(s) => PrimitiveValue::String(String::from_str_in(&s, &env.allocator)),
        _ => todo!("Lower literal {literal:#?}"),
    }
}
