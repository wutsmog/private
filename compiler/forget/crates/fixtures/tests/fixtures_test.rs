use build_hir::build;
use bumpalo::Bump;
use estree::{ModuleItem, Statement};
use estree_swc::parse;
use hir::{Environment, Print, Registry};
use insta::{assert_snapshot, glob};

#[test]
fn fixtures() {
    glob!("fixtures/**.js", |path| {
        let input = std::fs::read_to_string(path).unwrap();
        let ast = parse(&input, path.to_str().unwrap()).unwrap();

        let mut output = String::new();

        for (ix, item) in ast.body.into_iter().enumerate() {
            if let ModuleItem::Statement(stmt) = item {
                if let Statement::FunctionDeclaration(fun) = *stmt {
                    let allocator = Bump::new();
                    let environment = allocator.alloc(Environment::new(
                        &allocator,
                        hir::Features {
                            validate_frozen_lambdas: true,
                        },
                        Registry,
                    ));
                    let hir = build(&environment, *fun).unwrap();

                    if ix != 0 {
                        output.push_str("\n\n");
                    }
                    hir.print(&mut output).unwrap();
                }
            }
        }

        assert_snapshot!(format!("Input:\n{input}\n\nOutput:\n{output}"));
    });
}
