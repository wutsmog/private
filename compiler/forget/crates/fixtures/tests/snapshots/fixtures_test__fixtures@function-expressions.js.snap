---
source: crates/fixtures/tests/fixtures_test.rs
expression: "format!(\"Input:\\n{input}\\n\\nOutput:\\n{output}\")"
input_file: crates/fixtures/tests/fixtures/function-expressions.js
---
Input:
function Component(props) {
  const y = 2;
  const foo = function foo(x) {
    let a = 1;
    let b;
    if (a === 1) {
      b = 5 + 3;
    } else {
      b = false;
    }
    return x + y + b;
  };
}


Output:
function Component(
  unknown props$6,
)
entry bb0
bb0 (block)
  [0] #0 = 2
  [1] #1 = StoreLocal Const unknown y$7 = unknown #0
  [2] #2 = Function @deps[] @context[]:
      function foo(
        unknown x$8,
      )
      entry bb1
      bb1 (block)
        [0] #0 = 1
        [1] #1 = StoreLocal Let unknown a$9 = unknown #0
        [2] #2 = DeclareLocal Let unknown b$10
        [3] #9 = 1
        [4] #10 = 1
        [5] #11 = true
        [6] Goto bb3
      bb3 (block)
        predecessors: bb1
        [7] #3 = 5
        [8] #4 = 3
        [9] #5 = 8
        [10] #6 = StoreLocal Reassign unknown b$11 = unknown #5
        [11] Goto bb2
      bb2 (block)
        predecessors: bb3
        [12] #12 = LoadLocal unknown x$8
        [13] #13 = LoadLocal unknown y$0
        [14] #14 = Binary unknown #12 + unknown #13
        [15] #15 = 8
        [16] #16 = Binary unknown #14 + unknown #15
        [17] Return unknown #16
  [3] #3 = StoreLocal Const unknown foo$16 = unknown #2
  [4] #4 = <undefined>
  [5] Return unknown #4
